File: src/test/java/com/zepben/resultscache/blobstore/BlobStoreResultsCacheTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.resultscache.blobstore;

import com.zepben.blobstore.BlobReader;
import com.zepben.blobstore.BlobStore;
import com.zepben.blobstore.BlobWriter;
import com.zepben.resultscache.ResultsCache;
import com.zepben.resultscache.ResultsCacheException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;

import java.math.BigInteger;
import java.time.Duration;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

import static com.zepben.testutils.exception.ExpectException.expect;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.greaterThan;
import static org.hamcrest.Matchers.lessThanOrEqualTo;
import static org.hamcrest.core.IsNot.not;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

public class BlobStoreResultsCacheTest {

    private static final int NUM_VALUES_TO_ADD = 1000;

    private BlobStore blobStore = mock(BlobStore.class);
    private BlobReader blobReader = mock(BlobReader.class);
    private BlobWriter blobWriter = mock(BlobWriter.class);

    private ResultsCache cache;

    @BeforeEach
    public void before() {
        when(blobStore.reader()).thenReturn(blobReader);
        when(blobStore.writer()).thenReturn(blobWriter);
        cache = new BlobStoreResultsCache(blobStore);
    }

    @Test
    public void works() throws Exception {
        Map<String, byte[]> values = new HashMap<>();
        final int[] putCount = {0};

        when(blobReader.get(ArgumentMatchers.<String>any(), any()))
            .then(invocation -> values.get(invocation.getArgument(0).toString()));
        when(blobWriter.write(any(), any(), any()))
            .then(invocation -> {
                values.put(invocation.getArgument(0).toString(), invocation.getArgument(2));
                ++putCount[0];
                return true;
            });

        Random random = new Random();
        for (int i = 0; i < NUM_VALUES_TO_ADD; ++i) {
            byte[] value = new BigInteger(160, random).toString(32).getBytes();
            assertThat(value, notNullValue());
            assertThat(value.length, greaterThan(0));

            String key = cache.put(value);
            assertThat(key, notNullValue());
            assertThat(key, not(equalTo("")));
            assertThat(cache.get(key), equalTo(value));
        }
        assertThat(values.size(), equalTo(NUM_VALUES_TO_ADD));
        verify(blobWriter, times(putCount[0])).commit();

        cache.close();
        expect(() -> cache.get("key")).toThrow(ResultsCacheException.class).withMessage("Results cache has been closed");
        expect(() -> cache.put("value".getBytes())).toThrow(ResultsCacheException.class).withMessage("Results cache has been closed");
        expect(() -> cache.addTimeToLive("key")).toThrow(ResultsCacheException.class).withMessage("Results cache has been closed");
        expect(() -> cache.updateTimeToLive("key")).toThrow(ResultsCacheException.class).withMessage("Results cache has been closed");
        expect(() -> cache.processTimeToLive(Duration.ZERO)).toThrow(ResultsCacheException.class).withMessage("Results cache has been closed");
        cache.close();
    }

    @Test
    public void addTimeToLive() throws Exception {
        String key = "key";
        cache.addTimeToLive(key);

        ArgumentCaptor<byte[]> bytesCaptor = ArgumentCaptor.forClass(byte[].class);
        verify(blobWriter).write(eq(key), eq(BlobStoreResultsCache.TTL_ATTR), bytesCaptor.capture());
        verify(blobWriter).commit();

        assertInstantIsCloseToNow(decodeInstant(bytesCaptor.getValue()));
    }

    @Test
    public void updateTimeToLive() throws Exception {
        String key = "key";
        cache.updateTimeToLive(key);

        ArgumentCaptor<byte[]> bytesCaptor = ArgumentCaptor.forClass(byte[].class);
        verify(blobWriter).update(eq(key), eq(BlobStoreResultsCache.TTL_ATTR), bytesCaptor.capture());
        verify(blobWriter).commit();

        assertInstantIsCloseToNow(decodeInstant(bytesCaptor.getValue()));
    }

    @Test
    public void processesTtl() throws Exception {
        String keepId = "keep";
        String deleteId = "delete";

        Instant now = Instant.now();
        Map<String, byte[]> values = new HashMap<>();
        values.put(keepId, now.plusSeconds(5).toString().getBytes(BlobStoreResultsCache.BYTE_ENCODING));
        values.put(deleteId, now.minusSeconds(5).toString().getBytes(BlobStoreResultsCache.BYTE_ENCODING));

        doReturn(values).when(blobReader).getAll(BlobStoreResultsCache.TTL_ATTR);
        cache.processTimeToLive(Duration.ofSeconds(1));

        verify(blobReader).getAll(BlobStoreResultsCache.TTL_ATTR);
        verify(blobWriter, never()).delete(keepId);
        verify(blobWriter).delete(deleteId);
        verify(blobWriter).commit();
    }

    private Instant decodeInstant(byte[] instantBytes) {
        String timeStr = new String(instantBytes, BlobStoreResultsCache.BYTE_ENCODING);
        return Instant.parse(timeStr);
    }

    private void assertInstantIsCloseToNow(Instant instant) {
        long secondsAgo = Instant.now().getEpochSecond() - instant.getEpochSecond();
        assertThat(secondsAgo, lessThanOrEqualTo(1L));
    }

}

File: src/test/java/com/zepben/resultscache/blobstore/BlobStoreResultsCacheTest.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.resultscache.blobstore

import com.zepben.blobstore.BlobReader
import com.zepben.blobstore.BlobStore
import com.zepben.blobstore.BlobWriter
import com.zepben.resultscache.ResultsCacheException
import com.zepben.resultscache.blobstore.BlobStoreResultsCache.Companion.BYTE_ENCODING
import com.zepben.testutils.exception.ExpectException.Companion.expect
import io.mockk.*
import org.hamcrest.MatcherAssert.*
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.math.BigInteger
import java.time.Duration
import java.time.Instant
import java.util.*

class BlobStoreResultsCacheTest {

    private val blobReader = mockk<BlobReader>(relaxed = true)
    private val blobWriter = mockk<BlobWriter>(relaxed = true)

    private val blobStore = mockk<BlobStore> {
        every { reader() } returns blobReader
        every { writer() } returns blobWriter
        justRun { close() }
    }

    private val cache = BlobStoreResultsCache(blobStore)

    @Test
    fun works() {
        val values = mutableMapOf<String, ByteArray>()
        var putCount = 0

        every { blobReader[any<String>(), any()] } answers { values[firstArg()] }
        every { blobWriter.write(any(), any(), any()) } answers {
            values[firstArg()] = thirdArg()
            ++putCount
            true
        }

        val random = Random()
        for (i in 0 until NUM_VALUES_TO_ADD) {
            val value = BigInteger(160, random).toString(32).toByteArray(BYTE_ENCODING)
            assertThat(value, notNullValue())
            assertThat(value.size, greaterThan(0))

            val key = cache.put(value)
            assertThat(key, notNullValue())
            assertThat(key, not(equalTo("")))
            assertThat(cache[key], equalTo(value))
        }

        assertThat(values.size, equalTo(NUM_VALUES_TO_ADD))

        verify(exactly = putCount) { blobWriter.commit() }

        cache.close()

        expect { cache["key"] }.toThrow<ResultsCacheException>().withMessage("Results cache has been closed")
        expect { cache.put("value".toByteArray(BYTE_ENCODING)) }.toThrow<ResultsCacheException>().withMessage("Results cache has been closed")
        expect { cache.addTimeToLive("key") }.toThrow<ResultsCacheException>().withMessage("Results cache has been closed")
        expect { cache.updateTimeToLive("key") }.toThrow<ResultsCacheException>().withMessage("Results cache has been closed")
        expect { cache.processTimeToLive(Duration.ZERO) }.toThrow<ResultsCacheException>().withMessage("Results cache has been closed")

        // Make sure we can call close again without an exception.
        cache.close()

        // Make sure we only closed the store once.
        verify(exactly = 1) { blobStore.close() }
    }

    @Test
    @Throws(Exception::class)
    fun addTimeToLive() {
        val key = "key"
        cache.addTimeToLive(key)

        val bytesCaptor = slot<ByteArray>()
        verify { blobWriter.write(eq(key), eq(BlobStoreResultsCache.TTL_ATTR), capture(bytesCaptor)) }
        verify { blobWriter.commit() }

        assertInstantIsCloseToNow(decodeInstant(bytesCaptor.captured))
    }

    @Test
    @Throws(Exception::class)
    fun updateTimeToLive() {
        val key = "key"
        cache.updateTimeToLive(key)

        val bytesCaptor = slot<ByteArray>()
        verify { blobWriter.update(eq(key), eq(BlobStoreResultsCache.TTL_ATTR), capture(bytesCaptor)) }
        verify { blobWriter.commit() }

        assertInstantIsCloseToNow(decodeInstant(bytesCaptor.captured))
    }

    @Test
    @Throws(Exception::class)
    fun processesTtl() {
        val keepId = "keep"
        val deleteId = "delete"

        val now = Instant.now()
        val values: MutableMap<String, ByteArray> = HashMap()
        values[keepId] = now.plusSeconds(5).toString().toByteArray(BYTE_ENCODING)
        values[deleteId] = now.minusSeconds(5).toString().toByteArray(BYTE_ENCODING)

        every { blobReader.getAll(BlobStoreResultsCache.TTL_ATTR) } returns values
        cache.processTimeToLive(Duration.ofSeconds(1))

        verify { blobReader.getAll(BlobStoreResultsCache.TTL_ATTR) }
        verify(exactly = 0) { blobWriter.delete(keepId) }
        verify { blobWriter.delete(deleteId) }
        verify { blobWriter.commit() }
    }

    private fun decodeInstant(instantBytes: ByteArray): Instant {
        val timeStr = String(instantBytes, BYTE_ENCODING)
        return Instant.parse(timeStr)
    }

    private fun assertInstantIsCloseToNow(instant: Instant) {
        val secondsAgo = Instant.now().epochSecond - instant.epochSecond
        assertThat(secondsAgo, lessThanOrEqualTo(1L))
    }

    companion object {

        private const val NUM_VALUES_TO_ADD = 1000

    }

}
